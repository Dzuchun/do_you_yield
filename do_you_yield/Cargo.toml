[package]
name = "do_you_yield"
version = "0.1.0"
edition.workspace = true

[dev-dependencies]
trybuild = "1.0"

[dependencies]
do_you_yield_macro.workspace = true
