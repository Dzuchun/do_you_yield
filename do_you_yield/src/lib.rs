#![cfg_attr(not(test), no_std)]

pub use do_you_yield_macro::gn;

#[doc(hidden)]
pub mod sync;

#[doc(hidden)]
#[cfg(feature = "async")]
pub mod not_sync;

mod waker;
