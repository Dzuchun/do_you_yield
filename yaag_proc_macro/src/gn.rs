use crate::trans::Trans;
use syn::{Block, Expr, Ident, Token, Type, parse::Parse, parse_quote};

pub struct Input {
    is_async: bool,
    is_move: bool,
    code: Block,
    out: Type,
}

impl Parse for Input {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut is_async = false;
        let mut is_move = false;
        while !input.peek(Ident) {
            if input.parse::<Token![async]>().is_ok() {
                is_async = true;
                continue;
            }
            if input.parse::<Token![move]>().is_ok() {
                is_move = true;
                continue;
            }
            return Err(input.error("Expected move or async"));
        }
        let r#gen = input.parse::<Ident>().unwrap();
        if r#gen != "gen" {
            return Err(syn::Error::new_spanned(r#gen, "Expected `gen` keyword"));
        }
        let code = input.parse::<Block>()?;
        let _ = input.parse::<Token![->]>()?;
        let out = input.parse::<Type>()?;
        if !input.is_empty() {
            return Err(input.error("Extra input"));
        }
        Ok(Self {
            is_async,
            is_move,
            code,
            out,
        })
    }
}

impl Input {
    pub fn expand(mut self) -> Expr {
        let mv: Option<Token![move]> = self.is_move.then(Default::default);
        let out = self.out;
        self.code.trans(&out, self.is_async);
        let code = self.code;
        let module: Ident = if self.is_async {
            parse_quote!(not_sync)
        } else {
            parse_quote!(sync)
        };
        parse_quote! {{
            #[allow(unused_unsafe)]
            let fut = async #mv #code;
            ::yaag::#module::Gn {
                fut,
                _ph: ::core::marker::PhantomData::<#out>,
            }
        }}
    }
}
