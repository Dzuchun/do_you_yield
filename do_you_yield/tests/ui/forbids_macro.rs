use do_you_yield::gn;

fn main() {
    gn! {
        gen {
            let int = 42;
            let _int = core::pin::pin!(int);
        } -> i32
    };
}
