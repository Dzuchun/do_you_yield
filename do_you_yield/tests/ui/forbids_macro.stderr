error:
       Macro invocations are not allowed inside of the main generator code. Please do one of the following:
       - extract it into a function
       - extract it into a closure
       - move it out of the generator
 --> tests/ui/forbids_macro.rs:7:24
  |
7 |             let _int = core::pin::pin!(int);
  |                        ^^^^
